File: payharness/packages/sdk-python/payharness/webhook.py
import hashlib
import hmac
import time

TOLERANCE_SECONDS = 300


def create_webhook_signature(secret, timestamp, raw_body):
    key = hashlib.sha256(secret.encode()).digest()
    message = f"{timestamp}.{raw_body}".encode()
    digest = hmac.new(key, message, hashlib.sha256).hexdigest()
    return f"t={timestamp},v1={digest}"


def verify_webhook_signature(secret, signature_header, raw_body, now_seconds=None, tolerance_seconds=TOLERANCE_SECONDS):
    parts = signature_header.strip().split(",")
    if len(parts) != 2 or not parts[0].startswith("t=") or not parts[1].startswith("v1="):
        return False
    try:
        timestamp = int(parts[0][2:])
    except ValueError:
        return False
    digest = parts[1][3:]
    if len(digest) != 64 or any(c not in "0123456789abcdef" for c in digest):
        return False
    now = int(time.time()) if now_seconds is None else now_seconds
    if abs(now - timestamp) > tolerance_seconds:
        return False
    expected = create_webhook_signature(secret, timestamp, raw_body).split(",", 1)[1][3:]
    return hmac.compare_digest(expected, digest)

File: payharness/packages/sdk-go/webhook.go
package payharness

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"strconv"
	"strings"
	"time"
)

const WebhookTolerance = 5 * time.Minute

// CreateWebhookSignature creates a PayHarness webhook signature.
func CreateWebhookSignature(secret string, timestamp int64, rawBody string) string {
	keyHash := sha256.Sum256([]byte(secret))
	message := strconv.FormatInt(timestamp, 10) + "." + rawBody
	mac := hmac.New(sha256.New, keyHash[:])
	_, _ = mac.Write([]byte(message))
	return "t=" + strconv.FormatInt(timestamp, 10) + ",v1=" + hex.EncodeToString(mac.Sum(nil))
}

// VerifyWebhookSignature verifies a signature and rejects timestamps outside the tolerance window.
func VerifyWebhookSignature(secret, signature, rawBody string, now time.Time) bool {
	parts := strings.Split(signature, ",")
	if len(parts) != 2 || !strings.HasPrefix(parts[0], "t=") || !strings.HasPrefix(parts[1], "v1=") {
		return false
	}

	timestamp, err := strconv.ParseInt(strings.TrimPrefix(parts[0], "t="), 10, 64)
	if err != nil {
		return false
	}
	digest, err := hex.DecodeString(strings.TrimPrefix(parts[1], "v1="))
	if err != nil || len(digest) != sha256.Size {
		return false
	}

	when := time.Unix(timestamp, 0)
	if now.Sub(when) > WebhookTolerance || when.Sub(now) > WebhookTolerance {
		return false
	}

	expected := CreateWebhookSignature(secret, timestamp, rawBody)
	expectedDigest, _ := hex.DecodeString(strings.SplitN(expected, ",v1=", 2)[1])
	return hmac.Equal(expectedDigest, digest)
}
